server/parsec/components/memory/shamir.py

```diff
@@ -0,0 +1,89 @@
+        self._data.organizations[organization_id].shamir_setup.pop(author)
+
+    async def add_recovery_setup(
+        self,
+        organization_id: OrganizationID,
+        author: UserID,
+        device: DeviceID,
+        author_verify_key: VerifyKey,
+        setup: authenticated_cmds.latest.shamir_recovery_setup.ShamirRecoverySetup,
+    ) -> None | ShamirAddRecoverySetupBadOutcome:
+        # organization exists and not expired
+        try:
+            org = self._data.organizations[organization_id]
+        except KeyError:
+            return ShamirAddRecoverySetupBadOutcome.ORGANIZATION_NOT_FOUND
+        if org.is_expired:
+            return ShamirAddRecoverySetupBadOutcome.ORGANIZATION_EXPIRED
+
+        # author exists and not revoked and not frozen
+        try:
+            author_user = org.users[device.user_id]
+        except KeyError:
+            return ShamirAddRecoverySetupBadOutcome.AUTHOR_NOT_FOUND
+        if author_user.is_revoked:
+            return ShamirAddRecoverySetupBadOutcome.AUTHOR_REVOKED
+        if author_user.is_frozen:
+            return ShamirAddRecoverySetupBadOutcome.AUTHOR_FROZEN
+
+        match verify_certificates(setup, device, author_verify_key):
+            case (brief, shares):
+                # all recipients exists and not revoked
+                for share_recipient in shares.keys():
+                    try:
+                        recipient_user = org.users[share_recipient]
+                    except KeyError:
+                        return ShamirAddRecoverySetupBadOutcome.INVALID_RECIPIENT
+                    if recipient_user.is_revoked:
+                        return ShamirAddRecoverySetupBadOutcome.INVALID_RECIPIENT
+                    if recipient_user.is_frozen:
+                        return ShamirAddRecoverySetupBadOutcome.INVALID_RECIPIENT
+
+                # check that certificate timestamps are strictly increasing in the shamir topic
+                match self._data.organizations[organization_id].last_shamir_certificate:
+                    case None:
+                        # no previous shamir setup
+                        pass
+                    case v if v < brief.timestamp:
+                        # previous shamir happened strictly before
+                        pass
+                    case _:
+                        return ShamirAddRecoverySetupBadOutcome.INVALID_CERTIFICATE
+
+                if self._data.organizations[organization_id].shamir_setup.get(author) is None:
+                    self._data.organizations[organization_id].shamir_setup[author] = (
+                        MemoryShamirSetup(setup.ciphered_data, setup.reveal_token, brief, shares)
+                    )
+                else:
+                    return ShamirAddRecoverySetupBadOutcome.ALREADY_SET
+
+            case VerifyCertificatesBadOutcome.INVALID_DATA:
+                return ShamirAddRecoverySetupBadOutcome.INVALID_CERTIFICATE
```

Comment:
Minor style comments: in the other methods, we tend to have a `match` to validate the certificate and then perform some logic once it's validated outside of the match case. This makes the flow of the method more obvious to the reader:
```python
        match verify_certificates(setup, device, author_verify_key):
            case (brief, shares):
                pass
            case VerifyCertificatesBadOutcome.INVALID_DATA:
                return ShamirAddRecoverySetupBadOutcome.INVALID_CERTIFICATE
        # Then do something with `brief` and `shares`
```

Reply:
and nor brief nor shares are scoped because python :snake: